src/SAWScript/Crucible/MIR/Override.hs

```diff
@@ -1204,48 +1204,61 @@ matchArg opts sc cc cs prepost md actual expectedTy expected =
-                  expectedRef
-
-          |  otherwise
-          -> fail_
+            (Ctx.Empty Ctx.:> Crucible.RV actualSliceRef Ctx.:> Crucible.RV actualSliceLenSym),
+     Mir.TyRef (Mir.TySlice _) _,
+     MS.SetupSlice slice)
+       | -- Currently, all slice lengths must be concrete, so the case below
+         -- should always succeed.
+         Just actualSliceLenBV <- W4.asBV actualSliceLenSym -> do
+         let actualSliceLen :: Int
+             actualSliceLen = fromInteger $ BV.asUnsigned actualSliceLenBV
+
+         let -- Retrieve the N in &[T; N], failing if the supplied type is
+             -- different.
+             arrRefTyLen :: Mir.Ty -> OverrideMatcher MIR w Int
+             arrRefTyLen (Mir.TyRef (Mir.TyArray _ len) _) = pure len
+             arrRefTyLen _ = fail_
+
+         let -- Take the actual slice value's underlying reference, convert it
+             -- to an array reference value, and match it against the expected
+             -- array reference value. See Note [Matching slices in overrides]
+             -- for why we do this.
+             matchSlice :: Mir.Ty -> SetupValue -> OverrideMatcher MIR w ()
+             matchSlice expectedArrRefTy expectedArrRef = do
+               arrLen <- arrRefTyLen expectedArrRefTy
+               Ctx.Empty Ctx.:> Crucible.RV actualArrRef Ctx.:> _ <-
+                 liftIO $ Mir.mirRef_peelIndexIO bak iTypes actualElemTpr actualSliceRef
+               let actualArrTy = Mir.TyArray actualElemTy arrLen
+               let actualArrTpr = Mir.MirVectorRepr actualElemTpr
+               let actualArrRefTy = Mir.TyRef actualArrTy actualMutbl
+               let actualArrRefShp = RefShape actualArrRefTy actualArrTy actualMutbl actualArrTpr
+               matchArg opts sc cc cs prepost md
+                 (MIRVal actualArrRefShp actualArrRef)
+                 expectedArrRefTy
+                 expectedArrRef
+
+         case slice of
+           MirSetupSliceRaw{} ->
+             panic "matchArg" ["SliceRaw not yet implemented"]
+           MirSetupSlice expectedArrRef -> do
+             -- Check that the length of the expected array reference value
+             -- matches that of the actual slice reference value...
+             expectedArrRefTy <- typeOfSetupValue cc tyenv nameEnv expectedArrRef
+             expectedSliceLen <- arrRefTyLen expectedArrRefTy
+             unless (expectedSliceLen == actualSliceLen) fail_
+             -- ...then match the reference values.
+             matchSlice expectedArrRefTy expectedArrRef
+           MirSetupSliceRange expectedArrRef expectedStart expectedEnd -> do
+             -- Check that the length of the range of the expected slice
+             -- reference value matches that of the actual slice reference
+             -- value...
+             expectedArrRefTy <- typeOfSetupValue cc tyenv nameEnv expectedArrRef
+             let expectedSliceLen = expectedEnd - expectedStart
```

Comment:
Shouldn't this check that ```expectedEnd``` is less than (or equal to) ```arrRefTyLen expectedArrRefTy```, at least as an assertion? (One would expect not to make it this far if it's not true, but it seems like a good thing to crosscheck.)

Reply:
A good question. In general, most of these style of checks are delegated to the code in `resolveSetupVal` (e.g., [here](https://github.com/GaloisInc/saw-script/blob/be24399f385acd966b1cb59331f636e8f1609047/src/SAWScript/Crucible/MIR/ResolveSetupValue.hs#L613-L620)), which is invoked on any value before it is passed as an argument to an override. That will ensure that a slice range value is consistent with its own underlying array's length. The checks here in `matchArg` are specifically for checking the values of two different slice values, which isn't checked as part of `resolveSetupVal`.

Admittedly, this separation of concerns isn't always adhered to in practice, as there are parts of `matchArg` that also do `resolveSetupVal`-style self-consistency checks. In the sake of doing the right thing in this specific example, I will:

1. Leave a comment here which states that the `expectedEnd <= arrRefTyLen expectedArrRefTy` check is done elsewhere, and
2. Add an expected-to-fail test case that ensures that this check actually occurs.

Reply:
Actually, there are already test cases for (2) in [`test_mir_verify_slices`](https://github.com/GaloisInc/saw-script/blob/be24399f385acd966b1cb59331f636e8f1609047/intTests/test_mir_verify_slices/test.saw#L32-L52).